src/backend/distributed/commands/multi_copy.c

```diff
@@ -1015,14 +1015,67 @@ CanUseBinaryCopyFormatForType(Oid typeId)
 		return false;
 	}
 
-	if (typeId >= FirstNormalObjectId)
+	/*
+	 * A row type could contain any types, possibly types that don't have
+	 * the binary input and output functions defined.
+	 *
+	 * TODO: Inspect the types inside the record and check if all of them
+	 * can be binary encoded. If so, it's safe to use binary encoding.
+	 */
+	if (type_is_rowtype(typeId))
 	{
-		char typeCategory = '\0';
-		bool typePreferred = false;
+		return false;
+	}
 
-		get_type_category_preferred(typeId, &typeCategory, &typePreferred);
-		if (typeCategory == TYPCATEGORY_ARRAY ||
-			typeCategory == TYPCATEGORY_COMPOSITE)
+	HeapTuple typeTup = typeidType(typeId);
+	Form_pg_type type = (Form_pg_type) GETSTRUCT(typeTup);
+	Oid elementType = type->typelem;
+#if PG_VERSION_NUM < PG_VERSION_14
+	char typeCategory = type->typcategory;
+#endif
```

Comment:
can be moved in the #if block below

Reply:
no, because it needs to be done before `ReleaseSysCache`